File: leetcode/0685-redundant-connection-ii/Solution.java
class Solution {
    public int[] findRedundantDirectedConnection(int[][] edges) {
        int n = edges.length;
        int[] nodeParent = new int[n + 1];
        int[] candidate1 = null;
        int[] candidate2 = null;
        int skipIndex = -1;

        for (int i = 0; i < n; i++) {
            int u = edges[i][0];
            int v = edges[i][1];
            if (nodeParent[v] != 0) {
                candidate1 = new int[]{nodeParent[v], v};
                candidate2 = new int[]{u, v};
                skipIndex = i;
            } else {
                nodeParent[v] = u;
            }
        }

        int[] ufParent = new int[n + 1];
        for (int i = 0; i <= n; i++) {
            ufParent[i] = i;
        }

        for (int i = 0; i < n; i++) {
            if (i == skipIndex) {
                continue;
            }
            int u = edges[i][0];
            int v = edges[i][1];
            int ru = find(ufParent, u);
            int rv = find(ufParent, v);
            if (ru == rv) {
                return candidate1 != null ? candidate1 : new int[]{u, v};
            }
            ufParent[ru] = rv;
        }

        return candidate2;
    }

    private int find(int[] parent, int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }
}

File: leetcode/0685-redundant-connection-ii/solution.py
from typing import List, Optional


class Solution:
    def findRedundantDirectedConnection(self, edges: List[List[int]]) -> List[int]:
        n = len(edges)
        node_parent = [0] * (n + 1)
        candidate1: Optional[List[int]] = None
        candidate2: Optional[List[int]] = None
        skip_index = -1

        for i, (u, v) in enumerate(edges):
            if node_parent[v] != 0:
                candidate1 = [node_parent[v], v]
                candidate2 = [u, v]
                skip_index = i
            else:
                node_parent[v] = u

        uf_parent = list(range(n + 1))

        def find(x: int) -> int:
            while uf_parent[x] != x:
                uf_parent[x] = uf_parent[uf_parent[x]]
                x = uf_parent[x]
            return x

        for i, (u, v) in enumerate(edges):
            if i == skip_index:
                continue
            ru, rv = find(u), find(v)
            if ru == rv:
                return candidate1 if candidate1 else [u, v]
            uf_parent[ru] = rv

        return candidate2
